crashtracker-ffi/src/collector_windows/api.rs

```diff
@@ -0,0 +1,737 @@
+        prefix: WER_CONTEXT_PREFIX,
+        ptr: static_slice.as_ptr(),
+        len,
+        suffix: WER_CONTEXT_SUFFIX,
+    };
+
+    Ok(wer_context)
+}
+
+// https://github.com/microsoft/win32metadata/issues/1044
+#[repr(align(16))]
+#[derive(Default)]
+struct AlignedContext {
+    ctx: CONTEXT,
+}
+
+unsafe fn walk_thread_stack(
+    process_handle: HANDLE,
+    thread_id: u32,
+    modules: &[ModuleInfo],
+) -> Result<StackTrace> {
+    let mut stacktrace = StackTrace::new_incomplete();
+    let thread_handle = OpenThread(THREAD_ALL_ACCESS, false, thread_id)?;
+    let mut context = AlignedContext::default();
+
+    #[cfg(target_arch = "x86_64")]
+    {
+        context.ctx.ContextFlags = CONTEXT_FULL_AMD64;
+    }
+    #[cfg(target_arch = "x86")]
+    {
+        context.ctx.ContextFlags = CONTEXT_FULL_X86;
+    }
+
+    GetThreadContext(thread_handle, &mut context.ctx)?;
+
+    let mut native_frame = STACKFRAME_EX::default();
+    let machine_type: u32;
+
+    #[cfg(target_arch = "x86_64")]
+    {
+        machine_type = IMAGE_FILE_MACHINE_AMD64.0 as u32;
+        native_frame.AddrPC.Offset = context.ctx.Rip;
+        native_frame.AddrPC.Mode = AddrModeFlat;
+        native_frame.AddrStack.Offset = context.ctx.Rsp;
+        native_frame.AddrStack.Mode = AddrModeFlat;
+        native_frame.AddrFrame.Offset = context.ctx.Rbp;
+        native_frame.AddrFrame.Mode = AddrModeFlat;
+    }
+
+    #[cfg(target_arch = "x86")]
+    {
+        machine_type = IMAGE_FILE_MACHINE_I386.0 as u32;
+        native_frame.AddrPC.Offset = context.ctx.Eip as u64;
+        native_frame.AddrPC.Mode = AddrModeFlat;
+        native_frame.AddrStack.Offset = context.ctx.Esp as u64;
+        native_frame.AddrStack.Mode = AddrModeFlat;
+        native_frame.AddrFrame.Offset = context.ctx.Ebp as u64;
+        native_frame.AddrFrame.Mode = AddrModeFlat;
+    }
+
+    loop {
+        let result = StackWalkEx(
+            machine_type,
+            process_handle,
+            thread_handle,
+            &mut native_frame,
+            &mut context as *mut _ as *mut c_void,
+            None,
+            None,
+            None,
+            None,
+            SYM_STKWALK_DEFAULT,
+        );
+
+        if result != TRUE {
+            break;
+        }
+
+        let mut frame = StackFrame::new();
+
+        frame.ip = Some(format!("{:x}", native_frame.AddrPC.Offset));
+        frame.sp = Some(format!("{:x}", native_frame.AddrStack.Offset));
+
+        // Find the module
+        let module = modules.iter().find(|module| {
+            module.base_address <= native_frame.AddrPC.Offset
+                && native_frame.AddrPC.Offset < module.base_address + module.size
+        });
+
+        if let Some(module) = module {
+            frame.module_base_address = Some(format!("{:x}", module.base_address));
+            frame.symbol_address = Some(format!(
+                "{:x}",
+                native_frame.AddrPC.Offset - module.base_address
```

Comment:
Is this guaranteed not to underflow?

Reply:
We could read garbage in the registers so nothing is guaranteed. Would Rust panic in case of underflow?